sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,458 @@
+      EXPECT_READ32(GPIO_DATA_IN_REG_OFFSET, reg_val);
+      bool pin_val = !exp_pin_val;
+      EXPECT_EQ(dif_gpio_pin_read(&gpio_, pin, &pin_val), kDifGpioOk);
+      EXPECT_EQ(pin_val, exp_pin_val);
+    }
+  }
+}
+
+// Write tests
+class WriteTest : public DifGpioTestInitialized {};
+
+TEST_F(WriteTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_all_write(nullptr, kAllOnes), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_pin_write(nullptr, 0, true), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_masked_write(nullptr, kAllOnes, kAllOnes), kDifGpioBadArg);
+}
+
+TEST_F(WriteTest, AllPins) {
+  const uint32_t val = 0xA5A5A5A5;
+  EXPECT_WRITE32(GPIO_DIRECT_OUT_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_all_write(&gpio_, val), kDifGpioOk);
+}
+
+// The GPIO device provides masked bit-level atomic writes to its DIRECT_OUT
+// and DIRECT_OE registers. A 32-bit mask and a 32-bit value
+//
+// mask = [mask_upper, mask_lower],
+// bits:   31......16  15.......0
+//
+// val = [val_upper, val_lower]
+// bits:  31.....16  15......0
+//
+// are written to the registers such that
+//
+// MASKED_*_UPPER = [mask_upper, val_upper]
+//            bits:  31......16  15......0
+//
+// MASKED_*_LOWER = [mask_lower, val_lower]
+//            bits:  31......16  15......0
+//
+// Functions that operate on individual pins basically work in the same way
+// after creating the equivalent mask and value.
+
+TEST_F(WriteTest, SinglePin) {
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 0, true), kDifGpioOk);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 15, false), kDifGpioOk);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 16, true), kDifGpioOk);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 31, false), kDifGpioOk);
+}
+
+TEST_F(WriteTest, Masked) {
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, 0xCDCD3322);
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, 0xABAB5544);
+  EXPECT_EQ(dif_gpio_masked_write(&gpio_, 0xABABCDCD, 0x55443322), kDifGpioOk);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, 0xABAB5544);
+  EXPECT_EQ(dif_gpio_masked_write(&gpio_, 0xABAB0000, 0x55443322), kDifGpioOk);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, 0xCDCD3322);
+  EXPECT_EQ(dif_gpio_masked_write(&gpio_, 0x0000CDCD, 0x55443322), kDifGpioOk);
+}
+
+// Output mode tests
+class OutputModeTest : public DifGpioTestInitialized {};
+
+TEST_F(OutputModeTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_output_mode_all_set(nullptr, kAllOnes), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_output_mode_pin_set(nullptr, 0, true), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(nullptr, kAllOnes, kAllOnes),
+            kDifGpioBadArg);
+}
+
+TEST_F(OutputModeTest, AllPins) {
+  const uint32_t val = 0xA5A5A5A5;
```

Comment:
constexpr... (not going to comment on the rest of the file for these, I think you get the point. =P)